### PR TITLE
boards/arm/samv7: Introduce common folder

### PR DESCRIPTION
**Pull request by nandojve**
## Summary

Add common folder to share common code between samv7 boards avoiding both duplicate and copy and paste practices.
This is a sequence with:

- Define internal memories sizes by SoC version to allow unify linker scripts using a template model.
- Move all scripts and tools to common folder.It adapts current boards to use the common infrastructure and linker script template.
- Move all MCUboot related code to common folder and drop all duplicated code and defines.
- Demonstrate how simple is add MCUboot support to samv71-xult board.

## Impact

samv7 boards.

## Testing

- PassCI
- samv7-xult

### CODE REVIEW

File: boards/arm/samv7/common/Kconfig [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	select MTD_BYTE_WRITE
+	select MTD_PARTITION
+	select MTD_PROGMEM
+	select MTD_PROGMEM_ERASESTATE
+
+menuconfig SAMV7_FORMAT_MCUBOOT
+	bool "MCUboot-bootable format"
+	default n
+	select ARCH_RAMFUNCS
+	select BOARDCTL_RESET
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:27]: This is not mandatory. I think we should remove it from here

File: boards/arm/samv7/common/Kconfig [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	select BOOT_MCUBOOT
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:28]: This is `app` option. We should remove it from here.

File: boards/arm/samv7/common/Kconfig [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	select SAMV7_PROGMEM
+	select SAMV7_PROGMEM_MCUBOOT_PARTITION
+	select SAMV7_SYSTEMRESET
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:31]: This is not mandatory for MCUboot. I think we should remove it

File: boards/arm/samv7/common/Kconfig [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+comment "MCUboot Application Image MCUBOOT Update support"
+
+config SAMV7_MCUBOOT_PRIMARY_SLOT_OFFSET
+	hex "MCUboot application image primary slot offset"
+	default "0x20000"
+
+config SAMV7_MCUBOOT_PRIMARY_SLOT_DEVPATH
+	string "Application image primary slot device path"
+	default "/dev/mcuboot0"
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:45]:
```suggestion
	default "/dev/ota0"
```

File: boards/arm/samv7/common/Kconfig [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+config SAMV7_MCUBOOT_SECONDARY_SLOT_OFFSET
+	hex "MCUboot application image secondary slot offset"
+	default "0x100000"
+
+config SAMV7_MCUBOOT_SECONDARY_SLOT_DEVPATH
+	string "Application image secondary slot device path"
+	default "/dev/mcuboot1"
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:53]:
```suggestion
	default "/dev/ota1"
```

File: boards/arm/samv7/common/Kconfig [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	hex "MCUboot application image slot size (in bytes)"
+	default "0xe0000"
+
+config SAMV7_MCUBOOT_SCRATCH_OFFSET
+	hex "MCUboot scratch partition offset"
+	default "0x1e0000"
+
+config SAMV7_MCUBOOT_SCRATCH_DEVPATH
+	string "Scratch partition device path"
+	default "/dev/mcubootS"
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:65]:
```suggestion
	default "/dev/otascratch"
```

File: boards/arm/samv7/common/Kconfig [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+config SAMV7_MCUBOOT_SCRATCH_SIZE
+	hex "MCUboot scratch partition size (in bytes)"
+	default "0x20000"
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:69]: Default should depend on `SAMV7_CHIP_MEM_FLASH_XXXX`

File: boards/arm/samv7/common/Kconfig [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	string "Application image secondary slot device path"
+	default "/dev/mcuboot1"
+
+config SAMV7_MCUBOOT_SLOT_SIZE
+	hex "MCUboot application image slot size (in bytes)"
+	default "0xe0000"
+
+config SAMV7_MCUBOOT_SCRATCH_OFFSET
+	hex "MCUboot scratch partition offset"
+	default "0x1e0000"
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:61]: Default should depend on `SAMV7_CHIP_MEM_FLASH_XXXX`

File: boards/arm/samv7/common/Kconfig [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	hex "MCUboot application image secondary slot offset"
+	default "0x100000"
+
+config SAMV7_MCUBOOT_SECONDARY_SLOT_DEVPATH
+	string "Application image secondary slot device path"
+	default "/dev/mcuboot1"
+
+config SAMV7_MCUBOOT_SLOT_SIZE
+	hex "MCUboot application image slot size (in bytes)"
+	default "0xe0000"
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:57]: Default should depend on `SAMV7_CHIP_MEM_FLASH_XXXX`

File: boards/arm/samv7/common/Kconfig [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+	hex "MCUboot application image primary slot offset"
+	default "0x20000"
+
+config SAMV7_MCUBOOT_PRIMARY_SLOT_DEVPATH
+	string "Application image primary slot device path"
+	default "/dev/mcuboot0"
+
+config SAMV7_MCUBOOT_SECONDARY_SLOT_OFFSET
+	hex "MCUboot application image secondary slot offset"
+	default "0x100000"
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:49]: Default should depend on `SAMV7_CHIP_MEM_FLASH_XXXX`

File: boards/arm/samv7/same70-qmtech/README.txt [hunk 11]
```diff
@@ -794,8 +794,8 @@ Configuration sub-directories
       CONFIG_MCUBOOT_BOOTLOADER=y
       CONFIG_MCUBOOT_ENABLE_LOGGING=y
 
-      CONFIG_SAME70QMTECH_FORMAT_MCUBOOT=y
-      CONFIG_SAME70QMTECH_MCUBOOT_BOOTLOADER=y
+      CONFIG_SAMV7_FORMAT_MCUBOOT=y
+      CONFIG_SAMV7_PROGMEM_NSECTORS=16
```
Comment (pkarashchenko) [inline boards/arm/samv7/same70-qmtech/README.txt:798]: Qmtech board has 4 NSECTORS -- that is a default value

File: boards/arm/samv7/same70-qmtech/README.txt [hunk 12]
```diff
@@ -815,5 +815,6 @@ Configuration sub-directories
       CONFIG_BOOT_MCUBOOT=y
       CONFIG_MCUBOOT_SLOT_CONFIRM_EXAMPLE=y
 
-      CONFIG_SAME70QMTECH_FORMAT_MCUBOOT=y
+      CONFIG_SAMV7_FORMAT_MCUBOOT=y
+      CONFIG_SAMV7_PROGMEM_NSECTORS=16
```
Comment (pkarashchenko) [inline boards/arm/samv7/same70-qmtech/README.txt:819]: Qmtech board has 4 NSECTORS -- that is a default value

File: boards/arm/samv7/common/Kconfig [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+	select ARCH_RAMFUNCS
+	select BOARDCTL_RESET
+	select BOOT_MCUBOOT
+	select SAMV7_PROGMEM
+	select SAMV7_PROGMEM_MCUBOOT_PARTITION
+	select SAMV7_SYSTEMRESET
+	---help---
+		The MCUboot support of loading the firmware images.
+
+if SAMV7_FORMAT_MCUBOOT
```
Comment (pkarashchenko) [inline boards/arm/samv7/common/Kconfig:35]: `SAMV7_MCUBOOT_BOOTLOADER` -- config option is missing

File: boards/arm/samv7/same70-qmtech/configs/mcuboot-confirm/defconfig [hunk 14]
```diff
@@ -43,12 +46,12 @@ CONFIG_RAM_SIZE=262144
...
 CONFIG_RAW_BINARY=y
 CONFIG_RR_INTERVAL=200
-CONFIG_SAME70QMTECH_FORMAT_MCUBOOT=y
+CONFIG_SAMV7_FORMAT_MCUBOOT=y
 CONFIG_SAMV7_GPIOA_IRQ=y
 CONFIG_SAMV7_GPIOD_IRQ=y
 CONFIG_SAMV7_GPIO_IRQ=y
 CONFIG_SAMV7_HSMCI0=y
-CONFIG_SAMV7_PROGMEM=y
+CONFIG_SAMV7_PROGMEM_NSECTORS=16
```
Comment (pkarashchenko) [inline boards/arm/samv7/same70-qmtech/configs/mcuboot-confirm/defconfig:54]:
```suggestion
```

File: boards/arm/samv7/same70-qmtech/configs/mcuboot-confirm/defconfig [hunk 15]
```diff
@@ -30,6 +30,9 @@ CONFIG_FAT_LCNAMES=y
 CONFIG_FAT_LFN=y
 CONFIG_FS_FAT=y
 CONFIG_FS_PROCFS=y
+CONFIG_MCUBOOT_PRIMARY_SLOT_PATH="/dev/mcuboot0"
+CONFIG_MCUBOOT_SCRATCH_PATH="/dev/mcubootS"
+CONFIG_MCUBOOT_SECONDARY_SLOT_PATH="/dev/mcuboot1"
```
Comment (pkarashchenko) [inline boards/arm/samv7/same70-qmtech/configs/mcuboot-confirm/defconfig:35]: Let's go with defaults: `/dev/ota0`, `/dev/ota1`, `/dev/otascratch`

File: boards/arm/samv7/same70-qmtech/configs/mcuboot-confirm/defconfig_debug [hunk 16]
```diff
@@ -41,6 +41,9 @@ CONFIG_FS_FAT=y
 CONFIG_FS_PROCFS=y
 CONFIG_I2CTOOL_MAXBUS=0
 CONFIG_INTELHEX_BINARY=y
+CONFIG_MCUBOOT_PRIMARY_SLOT_PATH="/dev/mcuboot0"
+CONFIG_MCUBOOT_SCRATCH_PATH="/dev/mcubootS"
+CONFIG_MCUBOOT_SECONDARY_SLOT_PATH="/dev/mcuboot1"
```
Comment (pkarashchenko) [inline boards/arm/samv7/same70-qmtech/configs/mcuboot-confirm/defconfig_debug:46]: This file is committed by mistake. We should remove it

File: boards/arm/samv7/same70-qmtech/configs/mcuboot-loader/defconfig [hunk 17]
```diff
@@ -32,6 +32,9 @@ CONFIG_FS_FAT=y
 CONFIG_FS_PROCFS=y
 CONFIG_MCUBOOT_BOOTLOADER=y
 CONFIG_MCUBOOT_ENABLE_LOGGING=y
+CONFIG_MCUBOOT_PRIMARY_SLOT_PATH="/dev/mcuboot0"
+CONFIG_MCUBOOT_SCRATCH_PATH="/dev/mcubootS"
+CONFIG_MCUBOOT_SECONDARY_SLOT_PATH="/dev/mcuboot1"
```
Comment (pkarashchenko) [inline boards/arm/samv7/same70-qmtech/configs/mcuboot-loader/defconfig:37]: Let's go with defaults: `/dev/ota0`, `/dev/ota1`, `/dev/otascratch`

File: boards/arm/samv7/same70-qmtech/configs/mcuboot-loader/defconfig [hunk 18]
```diff
@@ -44,13 +47,12 @@ CONFIG_RAM_SIZE=262144
...
-CONFIG_SAME70QMTECH_FORMAT_MCUBOOT=y
-CONFIG_SAME70QMTECH_MCUBOOT_BOOTLOADER=y
+CONFIG_SAMV7_FORMAT_MCUBOOT=y
 CONFIG_SAMV7_GPIOA_IRQ=y
-CONFIG_SAMV7_GPIOD_IRQ=y
+CONFIG_SAMV7_GPIOC_IRQ=y
 CONFIG_SAMV7_GPIO_IRQ=y
 CONFIG_SAMV7_HSMCI0=y
-CONFIG_SAMV7_PROGMEM=y
+CONFIG_SAMV7_PROGMEM_NSECTORS=16
```
Comment (pkarashchenko) [inline boards/arm/samv7/same70-qmtech/configs/mcuboot-loader/defconfig:55]:
```suggestion
```

File: boards/arm/samv7/same70-qmtech/configs/mcuboot-loader/defconfig [hunk 19]
```diff
@@ -44,13 +47,12 @@ CONFIG_RAM_SIZE=262144
 CONFIG_RAM_START=0x20400000
 CONFIG_RAW_BINARY=y
 CONFIG_RR_INTERVAL=200
-CONFIG_SAME70QMTECH_FORMAT_MCUBOOT=y
-CONFIG_SAME70QMTECH_MCUBOOT_BOOTLOADER=y
+CONFIG_SAMV7_FORMAT_MCUBOOT=y
```
Comment (pkarashchenko) [inline boards/arm/samv7/same70-qmtech/configs/mcuboot-loader/defconfig:50]: `CONFIG_SAMV7_MCUBOOT_BOOTLOADER` -- is missing

File: boards/arm/samv7/same70-qmtech/configs/mcuboot-loader/defconfig [hunk 20]
```diff
@@ -44,13 +47,12 @@ CONFIG_RAM_SIZE=262144
...
 CONFIG_SAMV7_GPIOA_IRQ=y
-CONFIG_SAMV7_GPIOD_IRQ=y
+CONFIG_SAMV7_GPIOC_IRQ=y
```
Comment (pkarashchenko) [inline boards/arm/samv7/same70-qmtech/configs/mcuboot-loader/defconfig:52]:
```suggestion
CONFIG_SAMV7_GPIOD_IRQ=y
```

**Review by pkarashchenko — CHANGES_REQUESTED** [review]

**Comment by nandojve** [comment]
Hi @pkarashchenko ,

Thank you for your comments.

I was thinking more about how integrade bootloaders into NuttX. Besides this PR helps to unify samv7 it still lacks for all NuttX platforms. If makes sense, I would like split this PR and keep only first 2 commits. The next step will be create a RFC to define all those variables that are required to allow an agnostic bootloader. I mean, there is no sense an App outside NuttX define things that are HW dependent. I think it should be the contrary. This way it will be possible that any external app pick configs from NuttX and configure itself which allows an agnostic bootloader.

If above makes sense I prefer split PR.

What do you think?

File: boards/arm/samv7/common/Kconfig (hunk 13 above)
Comment (nandojve) [reply boards/arm/samv7/common/Kconfig:35]: CONFIG_BOARDCTL_BOOT_IMAGE is selected when building the boot-loader and is enough.

File: boards/arm/samv7/same70-qmtech/configs/mcuboot-loader/defconfig (hunk 19 above)
Comment (nandojve) [reply boards/arm/samv7/same70-qmtech/configs/mcuboot-loader/defconfig:50]: CONFIG_BOARDCTL_BOOT_IMAGE have same function.

**Comment by nandojve** [comment]
I'm working to better integrate bootloaders into NuttX. To do that, it is necessary re-order commit sequence, patch MCUboot and update Apps. I'll open a new PR as draft following a RFC proposal.

**Pull request closed without merging**